MetaDataset.py

```diff
@@ -885,23 +884,140 @@ def init_seq_order(self, epoch=None, seq_list=None):
     """
     assert seq_list is None, "seq_list not supported for %s" % self.__class__
     need_reinit = self.epoch is None or self.epoch != epoch
-    super(CombinedDataset, self).init_seq_order(epoch=epoch, seq_list=seq_list)
-    self.rnd.seed(epoch or 1)
+    num_seqs_saved = self._num_seqs
+    super(CombinedDataset, self).init_seq_order(epoch=epoch, seq_list=seq_list)  # resets self._num_seqs
+    full_epoch = (epoch - 1) // self.partition_epoch + 1 if epoch else 1
+    self.rnd.seed(full_epoch)
 
     if not need_reinit:
+      self._num_seqs = num_seqs_saved
       return False
 
-    # First init sequence order for sub-datasets as usual to get a list of available sequences. This way sorting and
+    # First init sequence order for sub-datasets as usual to get a list of available sequences. This way, sorting and
     # partition epoch of the individual sub-datasets is still supported. Later we will call init_seq_order again with a
     # sequence list to e.g. apply joint sorting or partition epoch of all sequences.
     for dataset in self.datasets.values():
       dataset.init_seq_order(epoch=epoch)
 
-    self.dataset_sorted_seq_idx_list = []  # We will fill this as we go
-    self.used_num_seqs_per_subset = [0] * len(self.datasets)
+    try:
+      total_num_seqs = sum([self.datasets[k].num_seqs for k in sorted(self.datasets.keys())])
+    except Exception:
+      total_num_seqs = None
+
+    if total_num_seqs is not None:
+      self.used_num_seqs_per_subset = [
+        self.datasets[self.dataset_idx2key_map[dataset_idx]].num_seqs for dataset_idx in range(len(self.datasets))]
+      self.dataset_seq_idx_boundaries = self._create_dataset_seq_idx_boundaries()
+
+      if self.seq_ordering == "random_dataset":
+        seq_order = self._get_random_dataset_seq_order(total_num_seqs)
+      else:
+        seq_order = self.get_seq_order_for_epoch(
+          epoch=epoch, num_seqs=total_num_seqs, get_seq_len=self._get_seq_length)
+      self._num_seqs = len(seq_order)
+
+      # Apply seq_order to self.dataset_seq_idx.
+      # We have to re-calculate the seq_idx's because we will sort the datasets below.
+      self.dataset_sorted_seq_idx_list = []
+      counters = [0] * len(self.datasets)
+      for seq_idx in seq_order:
+        dataset_idx, _ = self._seq_idx_to_dataset_seq_idx(seq_idx)
+        dataset_seq_idx = counters[dataset_idx]
+        counters[dataset_idx] += 1
+        self.dataset_sorted_seq_idx_list.append((dataset_idx, dataset_seq_idx))
+
+      # We only want to load those sequences in the sub-datasets that appear in seq_order. For this, we extract
+      # sequence lists containing the subset of sequences for each dataset from seq_order.
+      seq_lists = [[] for _ in self.datasets]
+      for seq_idx in seq_order:
+        dataset_idx, dataset_seq_idx = self._seq_idx_to_dataset_seq_idx(seq_idx)
+        dataset = self.datasets[self.dataset_idx2key_map[dataset_idx]]
+        seq_tag = dataset.get_tag(dataset_seq_idx)
```

Comment:
I just noticed you call `get_tag` here, which is again not possible in general (unless you have loaded that seq). E.g. check the implementation `CachedDataset2.get_tag`.

There is `get_all_tags`,  which is for the whole corpus, i.e. using corpus seq idx.
(Unfortunately, the implementation in the base class `Dataset.get_all_tags` is pretty buggy, and has the same problem. We maybe should remove that, and also introduce sth like `have_all_tags`. But anyway, this is independent of your PR here...)

There is also `get_corpus_seq_idx`, which gets in the current epoch seq idx, and returns the corpus seq idx.

So some code like this could work:
```
seq_corpus_tags_per_ds = {}
for key, ds in self.datasets.items():
  assert ds.have_corpus_seq_idx()
  seq_corpus_tags_per_ds[key] = ds.get_all_tags()

...

for seq_idx in seq_order:
  ...
  dataset_corpus_seq_idx = dataset.get_corpus_seq_idx(dataset_seq_idx)
  seq_tag = seq_corpus_tags_per_ds[self.dataset_idx2key_map[dataset_idx]][dataset_corpus_seq_idx]
```

Reply:
`Dataset.get_all_tags()` looks strange. It claims to disable partition_epoch but it doesn't. And also it returns the tags in sorted order, therefore your code snippet does not work like this. If I use `seq_tag = seq_corpus_tags_per_ds[self.dataset_idx2key_map[dataset_idx]][dataset_seq_idx]` (so without calculating `dataset_corpus_seq_idx, instead using the sorted index) it works, but I assume this is not how `get_all_tags()` is supposed to be used.

As a separate issue, we had performance issues related to loading and passing tags when using a huge number of sequences. Therefore I would like to avoid loading a list of all tags into memory...

In our code, we even did not use tags here at all, instead I extended `Dataset.init_seq_order()` to take a list of "reinit_indices" that reorder the previously initialised sequence ordering. But that code is probably not general enough...

Reply:
> `Dataset.get_all_tags()` looks strange. It claims to disable partition_epoch but it doesn't.

Yes, that's what I meant by saying:

>> (Unfortunately, the implementation in the base class `Dataset.get_all_tags` is pretty buggy, and has the same problem. We maybe should remove that, and also introduce sth like have_all_tags. But anyway, this is independent of your PR here...)

So just ignore the `Dataset.get_all_tags` (base class) implementation for now.
(Or maybe fix it. Or maybe just throw `OptionalNotImplementError` in there.)

But still this API is valid, and as I described (corpus seq idx -> seq tag), and correctly implemented by many datasets.

> And also it returns the tags in sorted order, therefore your code snippet does not work like this.

This is just the buggy base class implementation. Ignore it.

> As a separate issue, we had performance issues related to loading and passing tags when using a huge number of sequences. Therefore I would like to avoid loading a list of all tags into memory...

It is usually anyway in memory already. But yes, this creates a new list (usually with references to already existing string objects).
Also yes, this is maybe not optimal.

However, if you don't like this API, then it means you also need to create a new API for this.
 
> In our code, we even did not use tags here at all, instead I extended `Dataset.init_seq_order()` to take a list of "reinit_indices" that reorder the previously initialised sequence ordering. But that code is probably not general enough...

This sounds very ugly, hacky, and not generic. We definitely should not do that.

Unless you maybe mean something different: The current `seq_list` argument is a list of seq tags. We could also allow it to be a list of corpus seq idx. This would be valid and clean and reasonable (for those datasets where corpus seq idx makes sense; basically where `have_corpus_seq_idx()` would return True).

However, we need to make sure that all datasets handle this correctly, or at least catch it and throw an error if not supported. You need to go through all existing dataset implementations and check that.